overlord/snapstate/handlers.go

```diff
@@ -1777,6 +1777,10 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) (err error) {
 		return err
 	}
 
+	if err := m.maybeDiscardNamespacesOnSnapdDowngrade(st, newInfo); err != nil {
+		return fmt.Errorf("cannot discard preserved namespaces: %v", err)
+	}
+
 	// Restore configuration of the target revision (if available) on revert
```

Comment:
shouldn't this happen instead before we link the snap?

Reply:
Moved up.